### ISSUE TITLE
Make C-extension threadsafe/Ractor-safe

### ISSUE THREAD

**Issue by mohamedhafez** [brianmario/mysql2#1374]
In planning ahead to the near future when TruffleRuby can run C-extensions marked with `rb_ext_ractor_safe(true)` in parallel, and for when Ractors are no longer just experimental, it would be great if the C-extension could be made threadsafe, or marked as such if it already is so!

**Comment by larskanis** [comment]
Ractor support is not just calling `rb_ext_ractor_safe(true)`. To make the library usable and safe in a Ractor context usually needs a bunch of changes. Lots of constants needs to be made shareable, objects which are not safely shareable needs to be marked as such, even if frozen and all shared internal data needs to be thread safe. Global states might be handled per-Ractor instead.

In ruby-pg that required a bunch of changes:

https://github.com/ged/ruby-pg/pull/519
https://github.com/ged/ruby-pg/pull/596
https://github.com/ged/ruby-pg/pull/636
https://github.com/ged/ruby-pg/pull/660
https://github.com/ged/ruby-pg/pull/674

Ruby's description of Racor safety is here: https://docs.ruby-lang.org/en/master/extension_rdoc.html#appendix-f-ractor-support

**Comment by mohamedhafez** [comment]
If there's a ton of work to be done and this is a "won't fix", that's totally understandable! I can close the issue in that case, especially since users who want Ractor-safety have trilogy as an alternative.

(But I did think it was worth asking: for plenty of the libraries I hit up, they already happened to be Ractor safe and a simple declaration was all that was required)
